pmd-core/src/main/java/net/sourceforge/pmd/internal/util/ClasspathClassLoader.java

```diff
@@ -233,12 +251,77 @@ public InputStream getResourceAsStream(String name) {
         return super.getResourceAsStream(name);
     }
 
+    private static InputStream newInputStreamFromJrtFilesystem(Path path) {
+        LOG.trace("Found {}", path);
+        try {
+            // Note: The input streams from JrtFileSystem are ByteArrayInputStreams and do not
+            // need to be closed - we don't need to track these. The filesystem itself needs to be closed at the end.
+            // See https://github.com/openjdk/jdk/blob/970cd202049f592946f9c1004ea92dbd58abf6fb/src/java.base/share/classes/jdk/internal/jrtfs/JrtFileSystem.java#L334
+            return Files.newInputStream(path);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static class ModuleFinder extends ClassVisitor {
+        private String moduleName;
+
+        protected ModuleFinder() {
+            super(Opcodes.ASM9);
+        }
+
+        @Override
+        public ModuleVisitor visitModule(String name, int access, String version) {
+            moduleName = name;
+            return null;
+        }
+
+        public String getModuleName() {
+            return moduleName;
+        }
+    }
+
+    private URL findModule(Enumeration<URL> moduleInfoUrls, String moduleName) throws IOException {
+        while (moduleInfoUrls.hasMoreElements()) {
+            URL url = moduleInfoUrls.nextElement();
+
+            ModuleFinder finder = new ModuleFinder();
+            try (InputStream inputStream = url.openStream()) {
+                ClassReader classReader = new ClassReader(inputStream);
+                classReader.accept(finder, ClassReader.SKIP_CODE | ClassReader.SKIP_DEBUG | ClassReader.SKIP_FRAMES);
+            }
+            if (moduleName.equals(finder.getModuleName())) {
+                return url;
+            }
+        }
+
+        return null;
+    }
+
     @Override
     public URL getResource(String name) {
         // Override to make it child-first. This is the method used by
         // pmd-java's type resolution to fetch classes, instead of loadClass.
         Objects.requireNonNull(name);
 
+        String moduleName = extractModuleName(name);
+        if (moduleName != null) {
+            try {
+                Enumeration<URL> moduleInfoUrls = findResources(MODULE_INFO_SUFFIX);
+                URL moduleUrl = findModule(moduleInfoUrls, moduleName);
+
+                // no match in this classloader, search in parents
+                if (moduleUrl == null) {
+                    moduleInfoUrls = getParent().getResources(MODULE_INFO_SUFFIX);
+                    moduleUrl = findModule(moduleInfoUrls, moduleName);
+                }
+
+                return moduleUrl;
```

Comment:
Should we keep a mapping `name -> URL` to cache the result? This whole branch seems like an expensive operation with a lot of IO.

Reply:
I'm keeping now a map. Right now, I don't think, it would really be necessary, as currently it is only triggered by `import module ...`. But probably this will be used more often in the future.